shared/src/main/kotlin/org/javacs/kt/classpath/CachedClassPathResolver.kt

```diff
@@ -0,0 +1,176 @@
+    var buildFileVersion by ClassPathMetadataCache.buildFileVersion
+}
+
+class ClassPathCacheEntryEntity(id: EntityID<Int>) : IntEntity(id) {
+    companion object : IntEntityClass<ClassPathCacheEntryEntity>(ClassPathCacheEntry)
+
+    var compiledJar by ClassPathCacheEntry.compiledJar
+    var sourceJar by ClassPathCacheEntry.sourceJar
+}
+
+class BuildScriptClassPathCacheEntryEntity(id: EntityID<Int>) : IntEntity(id) {
+    companion object : IntEntityClass<BuildScriptClassPathCacheEntryEntity>(BuildScriptClassPathCacheEntry)
+
+    var jar by BuildScriptClassPathCacheEntry.jar
+}
+
+/** A classpath resolver that caches another resolver */
+internal class CachedClassPathResolver(
+    private val wrapped: ClassPathResolver,
+    private val db: Database
+) : ClassPathResolver {
+    override val resolverType: String get() = "Cached + ${wrapped.resolverType}"
+
+    private var cachedClassPathEntries: Set<ClassPathEntry>
+        get() = transaction(db) {
+            ClassPathCacheEntryEntity.all().map {
+                ClassPathEntry(
+                    compiledJar = Paths.get(it.compiledJar),
+                    sourceJar = if (it.sourceJar != null) Paths.get(it.sourceJar) else null
+                )
+            }.toSet()
+        }
+        set(newEntries) = transaction(db) {
+            ClassPathCacheEntry.deleteAll()
+            newEntries.map {
+                ClassPathCacheEntryEntity.new {
+                    compiledJar = it.compiledJar.toString()
+                    sourceJar = it.sourceJar.toString()
+                }
+            }
+        }
+
+    private var cachedBuildScriptClassPathEntries: Set<Path>
+        get() = transaction(db) { BuildScriptClassPathCacheEntryEntity.all().map { Paths.get(it.jar) }.toSet() }
+        set(newEntries) = transaction(db) {
+            BuildScriptClassPathCacheEntry.deleteAll()
+            newEntries.map { BuildScriptClassPathCacheEntryEntity.new { jar = it.toString() } }
+        }
+
+    private var cachedClassPathMetadata
+        get() = transaction(db) {
+            ClassPathMetadataCacheEntity.all().map {
+                ClasspathMetadata(
+                    includesSources = it.includesSources,
+                    buildFileVersion = it.buildFileVersion
+                )
+            }.firstOrNull()
+        }
+        set(newClassPathMetadata) = transaction(db) {
+            ClassPathMetadataCache.deleteAll()
+            val newClassPathMetadataRow = newClassPathMetadata ?: ClasspathMetadata()
+            ClassPathMetadataCacheEntity.new {
+                includesSources = newClassPathMetadataRow.includesSources
+                buildFileVersion = newClassPathMetadataRow.buildFileVersion
+            }
+        }
+
+    init {
+        transaction(db) {
+            SchemaUtils.createMissingTablesAndColumns(
+                ClassPathMetadataCache, ClassPathCacheEntry, BuildScriptClassPathCacheEntry
+            )
```

Comment:
Is there a way for us to evolve the schema once this ships? Since we only use the db as a cache, perhaps we could just recreate all the tables if something mismatches with our current schema?

Reply:
That's more or less what I had in mind. `SchemaUtils.createMissingTablesAndColumns` is specifically designed to be a lightweight easy-to-use way to update the schema. For things that can't be handled with this, I think recreating everything is the way to go. I thought about using something like Flyway at some point, but that felt overkill for this purpose. Worst case scenario, the symbol index (and any other cached data) will need to be rebuilt when we make a schema change (but I don't see this happening very often, especially since only a few changes would actually require that).